--- a/base/trace_event/builtin_categories.h
+++ b/base/trace_event/builtin_categories.h
@@ -92,6 +92,8 @@ PERFETTO_DEFINE_CATEGORIES_IN_NAMESPACE_WITH_ATTRS(
     perfetto::Category("base.power").SetDescription(
       "Events about global system power and battery/thermal state.")
       .SetTags("toplevel"),
+    perfetto::Category("brave"),
+    perfetto::Category("brave.adblock"),
     perfetto::Category("browser").SetTags("navigation"),
     perfetto::Category("browsing_data"),
     perfetto::Category("CacheStorage"),
--- a/chrome/app/chrome_command_ids.h
+++ b/chrome/app/chrome_command_ids.h
@@ -116,6 +116,8 @@
 #define IDC_UNFOCUS_TAB_GROUP         34108
 
 // Page-related commands
+#define IDC_ADBLOCK_CONTEXT_BLOCK_ELEMENTS 35999
+
 #define IDC_BOOKMARK_THIS_TAB           35000
 #define IDC_BOOKMARK_ALL_TABS           35001
 #define IDC_VIEW_SOURCE                 35002
--- a/chrome/app/generated_resources.grd
+++ b/chrome/app/generated_resources.grd
@@ -603,6 +603,9 @@ are declared in tools/grit/grit_args.gni.
 
        <!-- content area context menus. Android does not use it -->
       <if expr="not is_android">
+        <message name="IDS_ADBLOCK_CONTEXT_BLOCK_ELEMENTS" desc="Message for context menu that enables the element picker UI to apply cosmetic filter rules to the page">
+          Block elements
+        </message>
         <message name="IDS_CONTENT_CONTEXT_INSPECTELEMENT" desc="The name of the Inspect Element command in the content area context menu">
           I&amp;nspect
         </message>
--- a/chrome/browser/BUILD.gn
+++ b/chrome/browser/BUILD.gn
@@ -341,6 +341,22 @@ source_set("command_updater_impl") {
   ]
 }
 
+source_set("cosmetic_filters") {
+  public = [
+    "cosmetic_filters/cosmetic_filters_tab_helper.h",
+  ]
+  sources = [
+    "cosmetic_filters/cosmetic_filters_tab_helper.cc",
+  ]
+  deps = [
+    "//base",
+    "//chrome/browser/ui",
+    "//chrome/browser/ui/browser_window",
+    "//components/cosmetic_filters/common:mojom",
+    "//third_party/blink/public/common",
+  ]
+}
+
 source_set("defaults") {
   public = [ "defaults.h" ]
   sources = [ "defaults.cc" ]
@@ -673,6 +689,7 @@ source_set("core") {
     ":chrome_content_browser_client",
     ":chrome_content_browser_client_parts",
     ":command_updater_impl",
+    ":cosmetic_filters",
     ":defaults",
     ":expired_flags_list",
     ":file_select_helper",
@@ -1617,6 +1634,7 @@ static_library("browser") {
     ":chrome_browser_main",
     ":chrome_content_browser_client",
     ":command_updater_impl",
+    ":cosmetic_filters",
     ":defaults",
     ":expired_flags_list",
     ":file_select_helper",
@@ -1682,6 +1700,8 @@ static_library("browser") {
     "//chrome/browser/bluetooth:impl",
     "//chrome/browser/bookmarks",
     "//chrome/browser/bookmarks:impl",
+    "//chrome/browser/brave_shields",
+    "//chrome/browser/brave_shields:brave_shields_impl",
     "//chrome/browser/breadcrumbs",
     "//chrome/browser/browsing_data:constants",
     "//chrome/browser/browsing_data:impl",
@@ -2048,11 +2068,16 @@ static_library("browser") {
     "//chrome/browser/webid:impl",
     "//components/activity_reporter",
     "//components/activity_reporter:buildflags",
+    "//components/brave_shields/core/common",
+    "//components/brave_shields/core/common:mojom",
     "//components/browser_apis/tab_strip:mojom",
     "//components/browser_apis/ui_controllers/toolbar:mojom",
     "//components/content_settings/browser/ui",
+    "//components/cosmetic_filters/browser",
+    "//components/cosmetic_filters/common:mojom",
     "//components/page_content_annotations/content",
     "//components/page_content_annotations/core",
+    "//components/strings:components_strings_grit",
     "//components/supervised_user/core/browser:device_parental_controls",
     "//components/supervised_user/core/browser:synthetic_field_trial_delegate",
     "//components/visibility_timer",
--- a/chrome/browser/browser_process.h
+++ b/chrome/browser/browser_process.h
@@ -92,6 +92,10 @@ namespace activity_reporter {
 class ActivityReporter;
 }
 
+namespace brave_shields {
+class AdBlockService;
+}
+
 namespace component_updater {
 class ComponentUpdateService;
 }
@@ -148,6 +152,8 @@ class BrowserProcess {
 
   virtual ~BrowserProcess();
 
+  virtual brave_shields::AdBlockService* ad_block_service() = 0;
+
   // Returns the UnownedUserDataHost associated with this browser process. This
   // is used to retrieve arbitrary features from the browser process without
   // requiring BrowserProcess to have knowledge of them.
--- a/chrome/browser/browser_process_impl.cc
+++ b/chrome/browser/browser_process_impl.cc
@@ -46,6 +46,7 @@
 #include "build/config/linux/dbus/buildflags.h"
 #include "chrome/browser/accessibility/soda_installer_impl.h"
 #include "chrome/browser/battery/battery_metrics.h"
+#include "chrome/browser/brave_shields/ad_block_subscription_download_manager_getter.h"
 #include "chrome/browser/browser_process_platform_part.h"
 #include "chrome/browser/buildflags.h"
 #include "chrome/browser/chrome_browser_main.h"
@@ -105,6 +106,9 @@
 #include "chrome/installer/util/google_update_settings.h"
 #include "components/activity_reporter/activity_reporter.h"
 #include "components/application_locale_storage/application_locale_storage.h"
+#include "components/brave_shields/content/browser/ad_block_service.h"
+#include "components/brave_shields/content/browser/ad_block_subscription_service_manager.h"
+#include "components/brave_shields/core/common/features.h"
 #include "components/breadcrumbs/core/application_breadcrumbs_logger.h"
 #include "components/breadcrumbs/core/breadcrumb_persistent_storage_util.h"
 #include "components/breadcrumbs/core/breadcrumbs_status.h"
@@ -741,6 +745,9 @@ void BrowserProcessImpl::PostDestroyThreads() {
 #endif
 
   features_->PostDestroyThreads();
+
+  // AdBlockService should outlive its own worker thread.
+  ad_block_service_.reset();
 }
 #endif  // !BUILDFLAG(IS_ANDROID)
 
@@ -1361,6 +1368,21 @@ StatusTray* BrowserProcessImpl::status_tray() {
   return status_tray_.get();
 }
 
+brave_shields::AdBlockService* BrowserProcessImpl::ad_block_service() {
+  if (!ad_block_service_) {
+    scoped_refptr<base::SequencedTaskRunner> task_runner(
+        base::ThreadPool::CreateSequencedTaskRunner(
+            {base::MayBlock(), base::TaskPriority::USER_BLOCKING,
+             base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN}));
+    ad_block_service_ = std::make_unique<brave_shields::AdBlockService>(
+        local_state(), GetApplicationLocale(), component_updater(), task_runner,
+        AdBlockSubscriptionDownloadManagerGetter(),
+        profile_manager()->user_data_dir().Append(
+            profile_manager()->GetInitialProfileDir()));
+  }
+  return ad_block_service_.get();
+}
+
 #if BUILDFLAG(SAFE_BROWSING_AVAILABLE)
 safe_browsing::SafeBrowsingService*
 BrowserProcessImpl::safe_browsing_service() {
--- a/chrome/browser/browser_process_impl.h
+++ b/chrome/browser/browser_process_impl.h
@@ -206,6 +206,7 @@ class BrowserProcessImpl : public BrowserProcess,
       std::unique_ptr<BackgroundModeManager> manager) override;
 #endif
   StatusTray* status_tray() override;
+  brave_shields::AdBlockService* ad_block_service() override;
 #if BUILDFLAG(SAFE_BROWSING_AVAILABLE)
   safe_browsing::SafeBrowsingService* safe_browsing_service() override;
 #endif
@@ -399,6 +400,8 @@ class BrowserProcessImpl : public BrowserProcess,
   std::unique_ptr<BackgroundModeManager> background_mode_manager_;
 #endif
 
+  std::unique_ptr<brave_shields::AdBlockService> ad_block_service_;
+
 #if BUILDFLAG(SAFE_BROWSING_AVAILABLE)
   bool created_safe_browsing_service_ = false;
   scoped_refptr<safe_browsing::SafeBrowsingService> safe_browsing_service_;
--- a/chrome/browser/chrome_content_browser_client.cc
+++ b/chrome/browser/chrome_content_browser_client.cc
@@ -116,6 +116,8 @@
 #include "chrome/browser/media/webrtc/webrtc_logging_controller.h"
 #include "chrome/browser/metrics/chrome_feature_list_creator.h"
 #include "chrome/browser/navigation_predictor/anchor_element_preloader.h"
+#include "chrome/browser/net/brave_proxying_url_loader_factory.h"
+#include "chrome/browser/net/brave_proxying_web_socket.h"
 #include "chrome/browser/net/chrome_network_delegate.h"
 #include "chrome/browser/net/profile_network_context_service.h"
 #include "chrome/browser/net/profile_network_context_service_factory.h"
@@ -226,6 +228,8 @@
 #include "components/autofill/core/common/autofill_features.h"
 #include "components/autofill/core/common/autofill_switches.h"
 #include "components/blocked_content/popup_blocker.h"
+#include "components/brave_shields/content/browser/ad_block_service.h"
+#include "components/brave_shields/content/browser/domain_block_navigation_throttle.h"
 #include "components/browsing_topics/browsing_topics_service.h"
 #include "components/captive_portal/core/buildflags.h"
 #include "components/content_settings/browser/page_specific_content_settings.h"
@@ -5652,6 +5656,21 @@ void ChromeContentBrowserClient::RemovePresentationObserver(
 void ChromeContentBrowserClient::CreateThrottlesForNavigation(
     content::NavigationThrottleRegistry& registry) {
   CreateAndAddChromeThrottlesForNavigation(registry);
+
+  content::NavigationHandle& navigation_handle = registry.GetNavigationHandle();
+  content::BrowserContext* context =
+      navigation_handle.GetWebContents()->GetBrowserContext();
+
+  // The HostContentSettingsMap might be null for some irregular profiles, e.g.
+  // the System Profile.
+  auto* host_content_settings_map =
+      HostContentSettingsMapFactory::GetForProfile(context);
+  if (host_content_settings_map) {
+    brave_shields::DomainBlockNavigationThrottle::MaybeCreateAndAdd(
+        registry, g_browser_process->ad_block_service(),
+        g_browser_process->ad_block_service()->custom_filters_provider(),
+        host_content_settings_map, g_browser_process->GetApplicationLocale());
+  }
 }
 
 std::vector<std::unique_ptr<content::CommitDeferringCondition>>
@@ -6692,6 +6711,9 @@ void ChromeContentBrowserClient::WillCreateURLLoaderFactory(
     bool* disable_secure_dns,
     network::mojom::URLLoaderFactoryOverridePtr* factory_override,
     scoped_refptr<base::SequencedTaskRunner> navigation_response_task_runner) {
+  BraveProxyingURLLoaderFactory<std::shared_ptr>::MaybeProxyRequest(
+      browser_context, frame, factory_builder, navigation_response_task_runner);
+
 #if BUILDFLAG(ENABLE_EXTENSIONS_CORE)
   auto* web_request_api =
       extensions::BrowserContextKeyedAPIFactory<extensions::WebRequestAPI>::Get(
@@ -6808,24 +6830,7 @@ ChromeContentBrowserClient::
 
 bool ChromeContentBrowserClient::WillInterceptWebSocket(
     content::RenderFrameHost* frame) {
-#if BUILDFLAG(ENABLE_EXTENSIONS_CORE)
-  if (!frame) {
-    return false;
-  }
-  const auto* web_request_api =
-      extensions::BrowserContextKeyedAPIFactory<extensions::WebRequestAPI>::Get(
-          frame->GetBrowserContext());
-
-  // NOTE: Some unit test environments do not initialize
-  // BrowserContextKeyedAPI factories for e.g. WebRequest.
-  if (!web_request_api) {
-    return false;
-  } else {
-    return web_request_api->MayHaveProxiesForFrame(frame);
-  }
-#else
-  return false;
-#endif
+  return (frame != nullptr);
 }
 
 void ChromeContentBrowserClient::CreateWebSocket(
@@ -6836,20 +6841,52 @@ void ChromeContentBrowserClient::CreateWebSocket(
     const std::optional<std::string>& user_agent,
     mojo::PendingRemote<network::mojom::WebSocketHandshakeClient>
         handshake_client) {
+  // PATCH NOTE: This emulates `BraveContentBrowserClient`, but because
+  // that is essentially a wrapper around this very class, this has to
+  // copy the old ChromeContentBrowserClient::WillInterceptWebSocket
+  // into a lambda here since the logic of that function has been replaced.
+  auto* proxy =
+      BraveProxyingWebSocket<std::shared_ptr>::ProxyWebSocket(
+          frame, std::move(factory), url, site_for_cookies, user_agent);
+
+  auto WillInterceptWebSocketImpl = [](content::RenderFrameHost* frame)-> bool {
 #if BUILDFLAG(ENABLE_EXTENSIONS_CORE)
-  // TODO(crbug.com/40195467): Request w/o a frame also should be proxied.
-  if (!frame) {
-    return;
-  }
-  auto* web_request_api =
-      extensions::BrowserContextKeyedAPIFactory<extensions::WebRequestAPI>::Get(
-          frame->GetBrowserContext());
+    if (!frame) {
+      return false;
+    }
+    const auto* web_request_api =
+        extensions::BrowserContextKeyedAPIFactory<extensions::WebRequestAPI>::Get(
+            frame->GetBrowserContext());
 
-  DCHECK(web_request_api);
-  web_request_api->ProxyWebSocket(frame, std::move(factory), url,
-                                  site_for_cookies, user_agent,
-                                  std::move(handshake_client));
+    // NOTE: Some unit test environments do not initialize
+    // BrowserContextKeyedAPI factories for e.g. WebRequest.
+    if (!web_request_api) {
+      return false;
+    } else {
+      return web_request_api->MayHaveProxiesForFrame(frame);
+    }
+#else
+    return false;
 #endif
+  };
+
+  if (WillInterceptWebSocketImpl(frame)) {
+#if BUILDFLAG(ENABLE_EXTENSIONS_CORE)
+    // TODO(crbug.com/40195467): Request w/o a frame also should be proxied.
+    if (!frame) {
+      return;
+    }
+    auto* web_request_api = extensions::BrowserContextKeyedAPIFactory<
+        extensions::WebRequestAPI>::Get(frame->GetBrowserContext());
+
+    DCHECK(web_request_api);
+    web_request_api->ProxyWebSocket(frame, std::move(factory), url,
+                                    site_for_cookies, user_agent,
+                                    std::move(handshake_client));
+#endif
+  } else {
+    proxy->Start(std::move(handshake_client));
+  }
 }
 
 void ChromeContentBrowserClient::WillCreateWebTransport(
--- a/chrome/browser/chrome_content_browser_client_receiver_bindings.cc
+++ b/chrome/browser/chrome_content_browser_client_receiver_bindings.cc
@@ -14,6 +14,7 @@
 #include "chrome/browser/chrome_content_browser_client.h"
 #include "chrome/browser/chrome_content_browser_client_parts.h"
 #include "chrome/browser/content_settings/content_settings_manager_delegate.h"
+#include "chrome/browser/cosmetic_filters/cosmetic_filters_tab_helper.h"
 #include "chrome/browser/headless/headless_mode_util.h"
 #include "chrome/browser/net/net_error_tab_helper.h"
 #include "chrome/browser/net_benchmarking.h"
@@ -27,7 +28,14 @@
 #include "chrome/common/buildflags.h"
 #include "chrome/common/chrome_features.h"
 #include "components/autofill/content/browser/content_autofill_driver_factory.h"
+#include "components/brave_shields/content/browser/ad_block_engine_wrapper.h"
+#include "components/brave_shields/content/browser/ad_block_service.h"
+#include "components/brave_shields/content/browser/domain_block_navigation_throttle.h"
+#include "components/brave_shields/core/common/brave_shield_constants.h"
+#include "components/brave_shields/core/common/features.h"
 #include "components/content_capture/browser/onscreen_content_provider.h"
+#include "components/cosmetic_filters/browser/cosmetic_filters_resources.h"
+#include "components/cosmetic_filters/common/cosmetic_filters.mojom.h"
 #include "components/metrics/call_stacks/call_stack_profile_collector.h"
 #include "components/offline_pages/buildflags/buildflags.h"
 #include "components/page_load_metrics/browser/metrics_web_contents_observer.h"
@@ -47,6 +55,7 @@
 #include "content/public/common/buildflags.h"
 #include "media/mojo/buildflags.h"
 #include "mojo/public/cpp/bindings/binder_map.h"
+#include "mojo/public/cpp/bindings/self_owned_receiver.h"
 #include "pdf/buildflags.h"
 #include "printing/buildflags/buildflags.h"
 #include "third_party/blink/public/common/associated_interfaces/associated_interface_registry.h"
@@ -203,6 +212,24 @@ void BindBadgeServiceForServiceWorker(
 }
 #endif
 
+void BindCosmeticFiltersResourcesOnTaskRunner(
+    mojo::PendingReceiver<cosmetic_filters::mojom::CosmeticFiltersResources>
+        receiver,
+    brave_shields::AdBlockEngineWrapper* engine_wrapper) {
+  mojo::MakeSelfOwnedReceiver(
+      std::make_unique<cosmetic_filters::CosmeticFiltersResources>(
+          engine_wrapper),
+      std::move(receiver));
+}
+
+void BindCosmeticFiltersResources(
+    content::RenderFrameHost* const frame_host,
+    mojo::PendingReceiver<cosmetic_filters::mojom::CosmeticFiltersResources>
+        receiver) {
+  g_browser_process->ad_block_service()->AsyncCall(base::BindOnce(
+      &BindCosmeticFiltersResourcesOnTaskRunner, std::move(receiver)));
+}
+
 }  // namespace
 
 void ChromeContentBrowserClient::ExposeInterfacesToRenderer(
@@ -317,6 +344,12 @@ void ChromeContentBrowserClient::BindMediaServiceReceiver(
 void ChromeContentBrowserClient::RegisterBrowserInterfaceBindersForFrame(
     content::RenderFrameHost* render_frame_host,
     mojo::BinderMapWithContext<content::RenderFrameHost*>* map) {
+  // NOTE: This is weird. brave registers this after calling this function,
+  // but doing this at the end instead means that something tries to message
+  // this before it's bound.
+  map->Add<cosmetic_filters::mojom::CosmeticFiltersResources>(
+      base::BindRepeating(&BindCosmeticFiltersResources));
+
   chrome::internal::PopulateChromeFrameBinders(map, render_frame_host);
   chrome::internal::PopulateChromeWebUIFrameBinders(map, render_frame_host);
 
@@ -394,6 +427,11 @@ void ChromeContentBrowserClient::
     RegisterAssociatedInterfaceBindersForRenderFrameHost(
         content::RenderFrameHost& render_frame_host,
         blink::AssociatedInterfaceRegistry& associated_registry) {
+  associated_registry.AddInterface<
+      cosmetic_filters::mojom::CosmeticFiltersHandler>(base::BindRepeating(
+      &cosmetic_filters::CosmeticFiltersTabHelper::BindCosmeticFiltersHandler,
+      &render_frame_host));
+
   for (auto& ep : extra_parts_) {
     ep->ExposeInterfacesToRendererForRenderFrameHost(render_frame_host,
                                                      associated_registry);
--- a/chrome/browser/download/background_download_service_factory.cc
+++ b/chrome/browser/download/background_download_service_factory.cc
@@ -9,6 +9,7 @@
 
 #include "base/feature_list.h"
 #include "base/files/file_path.h"
+#include "base/functional/bind.h"
 #include "base/memory/raw_ptr.h"
 #include "base/no_destructor.h"
 #include "base/task/sequenced_task_runner.h"
@@ -16,6 +17,7 @@
 #include "base/task/task_traits.h"
 #include "base/task/thread_pool.h"
 #include "build/build_config.h"
+#include "chrome/browser/browser_process.h"
 #include "chrome/browser/download/deferred_client_wrapper.h"
 #include "chrome/browser/download/download_manager_utils.h"
 #include "chrome/browser/download/simple_download_manager_coordinator_factory.h"
@@ -26,6 +28,9 @@
 #include "chrome/browser/transition_manager/full_browser_transition_manager.h"
 #include "chrome/common/chrome_constants.h"
 #include "components/background_fetch/download_client.h"
+#include "components/brave_shields/content/browser/ad_block_service.h"
+#include "components/brave_shields/content/browser/ad_block_subscription_download_client.h"
+#include "components/brave_shields/content/browser/ad_block_subscription_service_manager.h"
 #include "components/download/content/factory/download_service_factory_helper.h"
 #include "components/download/content/factory/navigation_monitor_factory.h"
 #include "components/download/public/background_service/background_download_service.h"
@@ -35,6 +40,7 @@
 #include "components/download/public/background_service/features.h"
 #include "components/download/public/background_service/url_loader_factory_getter.h"
 #include "components/download/public/common/simple_download_manager_coordinator.h"
+#include "components/download/public/task/task_scheduler.h"
 #include "components/keyed_service/core/simple_dependency_manager.h"
 #include "components/leveldb_proto/public/proto_database_provider.h"
 #include "components/optimization_guide/core/optimization_guide_features.h"
@@ -55,6 +61,17 @@
 
 namespace {
 
+brave_shields::AdBlockSubscriptionServiceManager*
+GetAdBlockSubscriptionServiceManager() {
+  return g_browser_process->ad_block_service()->subscription_service_manager();
+}
+
+std::unique_ptr<download::Client> CreateAdBlockSubscriptionDownloadClient(
+    Profile* profile) {
+  return std::make_unique<brave_shields::AdBlockSubscriptionDownloadClient>(
+      base::BindRepeating(&GetAdBlockSubscriptionServiceManager));
+}
+
 std::unique_ptr<download::Client> CreateBackgroundFetchDownloadClient(
     Profile* profile) {
   return std::make_unique<background_fetch::DownloadClient>(profile);
@@ -165,6 +182,11 @@ BackgroundDownloadServiceFactory::BuildServiceInstanceFor(
   auto clients = std::make_unique<download::DownloadClientMap>();
   ProfileKey* profile_key = ProfileKey::FromSimpleFactoryKey(key);
 
+  clients->insert(std::make_pair(
+      download::DownloadClient::CUSTOM_LIST_SUBSCRIPTIONS,
+      std::make_unique<download::DeferredClientWrapper>(
+          base::BindOnce(&CreateAdBlockSubscriptionDownloadClient), key)));
+
   clients->insert(std::make_pair(
       download::DownloadClient::BACKGROUND_FETCH,
       std::make_unique<download::DeferredClientWrapper>(
--- a/chrome/browser/net/BUILD.gn
+++ b/chrome/browser/net/BUILD.gn
@@ -26,6 +26,11 @@ proto_library("key_pinning_proto") {
 
 source_set("net") {
   public = [
+    "brave_ad_block_csp_network_delegate_helper.h",
+    "brave_ad_block_tp_network_delegate_helper.h",
+    "brave_proxying_url_loader_factory.h",
+    "brave_proxying_web_socket.h",
+    "brave_request_handler.h",
     "cert_verifier_service_time_updater.h",
     "chrome_mojo_proxy_resolver_factory.h",
     "chrome_network_delegate.h",
@@ -48,11 +53,13 @@ source_set("net") {
     "proxy_service_factory.h",
     "qwac_web_contents_observer.h",
     "referrer.h",
+    "resource_context_data.h",
     "secure_dns_config.h",
     "secure_dns_policy_handler.h",
     "secure_dns_util.h",
     "stub_resolver_config_reader.h",
     "system_network_context_manager.h",
+    "url_context.h",
   ]
 
   public_deps = [
@@ -120,6 +127,11 @@ source_set("net") {
 
 source_set("impl") {
   sources = [
+    "brave_ad_block_csp_network_delegate_helper.cc",
+    "brave_ad_block_tp_network_delegate_helper.cc",
+    "brave_proxying_url_loader_factory.cc",
+    "brave_proxying_web_socket.cc",
+    "brave_request_handler.cc",
     "cert_verifier_service_time_updater.cc",
     "chrome_mojo_proxy_resolver_factory.cc",
     "chrome_network_delegate.cc",
@@ -139,11 +151,13 @@ source_set("impl") {
     "proxy_service_factory.cc",
     "qwac_web_contents_observer.cc",
     "referrer.cc",
+    "resource_context_data.cc",
     "secure_dns_config.cc",
     "secure_dns_policy_handler.cc",
     "secure_dns_util.cc",
     "stub_resolver_config_reader.cc",
     "system_network_context_manager.cc",
+    "url_context.cc",
   ]
 
   public_deps = [
@@ -171,6 +185,9 @@ source_set("impl") {
     "//chrome/browser/ssl",
     "//chrome/browser/webid",
     "//chrome/common:version_header",
+    "//components/brave_shields/content/browser",
+    "//components/brave_shields/core/browser",
+    "//components/brave_shields/core/common",
     "//components/certificate_transparency",
     "//components/content_settings/core/common",
     "//components/country_codes",
--- a/chrome/browser/prefs/browser_prefs.cc
+++ b/chrome/browser/prefs/browser_prefs.cc
@@ -108,6 +108,7 @@
 #include "components/accessibility_annotator/core/prefs.h"
 #include "components/autofill/core/common/autofill_prefs.h"
 #include "components/blocked_content/safe_browsing_triggered_popup_blocker.h"
+#include "components/brave_shields/content/browser/ad_block_service.h"
 #include "components/breadcrumbs/core/breadcrumbs_status.h"
 #include "components/browsing_data/core/pref_names.h"
 #include "components/certificate_transparency/pref_names.h"
@@ -1322,6 +1323,7 @@ void RegisterLocalState(PrefRegistrySimple* registry) {
       registry);
 #endif
   autofill::prefs::RegisterLocalStatePrefs(registry);
+  brave_shields::RegisterPrefsForAdBlockService(registry);
   breadcrumbs::RegisterPrefs(registry);
   browser_shutdown::RegisterPrefs(registry);
   BrowserProcessImpl::RegisterPrefs(registry);
--- a/chrome/browser/profiles/chrome_browser_main_extra_parts_profiles.cc
+++ b/chrome/browser/profiles/chrome_browser_main_extra_parts_profiles.cc
@@ -48,6 +48,7 @@
 #include "chrome/browser/bitmap_fetcher/bitmap_fetcher_service_factory.h"
 #include "chrome/browser/bluetooth/bluetooth_chooser_context_factory.h"
 #include "chrome/browser/bookmarks/bookmark_model_factory.h"
+#include "chrome/browser/brave_shields/ad_block_pref_service_factory.h"
 #include "chrome/browser/breadcrumbs/breadcrumb_manager_keyed_service_factory.h"
 #include "chrome/browser/browser_process.h"
 #include "chrome/browser/browsing_data/browsing_data_history_observer_service.h"
@@ -1528,6 +1529,8 @@ void ChromeBrowserMainExtraPartsProfiles::
   autofill::AtMemoryPromoTrackerFactory::GetInstance();
 #endif
 
+  brave_shields::AdBlockPrefServiceFactory::GetInstance();
+
   WebDataServiceFactory::GetInstance();
   webrtc_event_logging::WebRtcEventLogManagerKeyedServiceFactory::GetInstance();
 }
--- a/chrome/browser/renderer_context_menu/render_view_context_menu.cc
+++ b/chrome/browser/renderer_context_menu/render_view_context_menu.cc
@@ -44,6 +44,7 @@
 #include "chrome/browser/autocomplete/autocomplete_classifier_factory.h"
 #include "chrome/browser/browser_features.h"
 #include "chrome/browser/browser_process.h"
+#include "chrome/browser/cosmetic_filters/cosmetic_filters_tab_helper.h"
 #include "chrome/browser/custom_handlers/protocol_handler_registry_factory.h"
 #include "chrome/browser/devtools/devtools_window.h"
 #include "chrome/browser/devtools/features.h"
@@ -160,6 +161,7 @@
 #include "components/autofill/core/common/autofill_features.h"
 #include "components/autofill/core/common/password_generation_util.h"
 #include "components/autofill/core/common/unique_ids.h"
+#include "components/brave_shields/core/common/features.h"
 #include "components/compose/buildflags.h"
 #include "components/compose/core/browser/compose_features.h"
 #include "components/contextual_tasks/public/features.h"
@@ -617,13 +619,14 @@ const std::map<int, int>& GetIdcToUmaMap(UmaEnumIdLookupType type) {
        {IDC_CONTENT_CONTEXT_LISTEN_TO_THIS_PAGE, 164},
        {IDC_CONTENT_CONTEXT_DICTATION, 165},
        {IDC_CONTENT_CONTEXT_ADD_LINK_TO_READING_LIST, 166},
+       {IDC_ADBLOCK_CONTEXT_BLOCK_ELEMENTS, 167},
        // To add new items:
        //   - Add one more line above this comment block, using the UMA value
        //     from the line below this comment block.
        //   - Increment the UMA value in that latter line.
        //   - Add the new item to the RenderViewContextMenuItem enum in
        //     tools/metrics/histograms/metadata/ui/enums.xml.
-       {0, 167}});
+       {0, 168}});
   // LINT.ThenChange(//tools/metrics/histograms/metadata/ui/enums.xml:RenderViewContextMenuItem)
 
   // LINT.IfChange(ContextMenuOptionDesktop)
@@ -1750,13 +1753,29 @@ void RenderViewContextMenu::AppendDeveloperItems() {
   bool hide_developer_items = IsDevToolsURL(params_.page_url) &&
                               params_.page_url.GetQuery().find(
                                   "debugFrontend=true") == std::string::npos;
-  if (hide_developer_items) {
+
+  bool add_block_elements =
+      (params_.selection_text.empty() || !params_.link_url.is_empty()) &&
+      params_.page_url.SchemeIsHTTPOrHTTPS() &&
+      base::FeatureList::IsEnabled(
+          brave_shields::features::kBraveShieldsElementPicker);
+
+  if (hide_developer_items && !add_block_elements) {
     return;
   }
 
   // In the DevTools popup menu, "developer items" is normally the only
   // section, so omit the separator there.
   menu_model_.AddSeparator(ui::NORMAL_SEPARATOR);
+
+  if (add_block_elements) {
+    menu_model_.AddItemWithStringId(IDC_ADBLOCK_CONTEXT_BLOCK_ELEMENTS,
+                                    IDS_ADBLOCK_CONTEXT_BLOCK_ELEMENTS);
+  }
+
+  if (hide_developer_items)
+    return;
+
   if (content_type_->SupportsGroup(ContextMenuContentType::ITEM_GROUP_PAGE)) {
     menu_model_.AddItemWithStringId(IDC_VIEW_SOURCE,
                                     IDS_CONTENT_CONTEXT_VIEWPAGESOURCE);
@@ -3281,6 +3300,9 @@ bool RenderViewContextMenu::IsCommandIdEnabled(int id) const {
   }
 #endif
   switch (id) {
+    case IDC_ADBLOCK_CONTEXT_BLOCK_ELEMENTS:
+      return true;
+
     case IDC_BACK:
       return embedder_web_contents_->GetController().CanGoBack();
 
@@ -3635,6 +3657,11 @@ void RenderViewContextMenu::ExecuteCommand(int id, int event_flags) {
   }
 
   switch (id) {
+    case IDC_ADBLOCK_CONTEXT_BLOCK_ELEMENTS:
+      cosmetic_filters::CosmeticFiltersTabHelper::LaunchContentPicker(
+          source_web_contents_);
+      break;
+
     case IDC_CONTENT_CONTEXT_OPENLINKNEWTAB: {
       WindowOpenDisposition new_tab_disposition =
           WindowOpenDisposition::NEW_BACKGROUND_TAB;
--- a/chrome/browser/resources/BUILD.gn
+++ b/chrome/browser/resources/BUILD.gn
@@ -20,6 +20,7 @@ assert(!is_fuchsia, "Fuchsia shouldn't use anything in //chrome")
 
 group("resources") {
   public_deps = [
+    "adblock_ui:resources",
     "contextual_tasks:resources",
     "glic:resources",
     "saved_tab_groups_unsupported:resources",
--- a/chrome/browser/ui/BUILD.gn
+++ b/chrome/browser/ui/BUILD.gn
@@ -104,6 +104,8 @@ static_library("ui") {
     "tab_ui_helper.h",
     "uninstall_browser_prompt.h",
     "view_ids.h",
+    "webui/brave_adblock_ui.cc",
+    "webui/brave_adblock_ui.h",
     "webui/chrome_web_ui_controller_factory.cc",
     "webui/chrome_web_ui_controller_factory.h",
     "webui/commerce/commerce_internals_ui_config.cc",
--- a/chrome/browser/ui/webui/chrome_web_ui_configs.cc
+++ b/chrome/browser/ui/webui/chrome_web_ui_configs.cc
@@ -95,6 +95,7 @@
 #include "chrome/browser/ui/webui/app_service_internals/app_service_internals_ui.h"
 #include "chrome/browser/ui/webui/autofill_ml_internals/autofill_ml_internals_ui.h"
 #include "chrome/browser/ui/webui/bookmarks/bookmarks_ui.h"
+#include "chrome/browser/ui/webui/brave_adblock_ui.h"
 #include "chrome/browser/ui/webui/color_pipeline_internals/color_pipeline_internals_ui.h"
 #include "chrome/browser/ui/webui/commerce/shopping_insights_side_panel_ui.h"
 #include "chrome/browser/ui/webui/downloads/downloads_ui.h"
@@ -252,6 +253,7 @@ void RegisterChromeWebUIConfigs() {
   map.AddWebUIConfig(std::make_unique<AccessibilityUIConfig>());
   map.AddWebUIConfig(std::make_unique<AutofillInternalsUIConfig>());
   map.AddWebUIConfig(std::make_unique<BluetoothInternalsUIConfig>());
+  map.AddWebUIConfig(std::make_unique<BraveAdblockUIConfig>());
   map.AddWebUIConfig(std::make_unique<BrowsingTopicsInternalsUIConfig>());
   map.AddWebUIConfig(
       std::make_unique<chrome_finds_internals::ChromeFindsInternalsUIConfig>());
--- a/chrome/chrome_paks.gni
+++ b/chrome/chrome_paks.gni
@@ -120,6 +120,7 @@ template("chrome_extra_paks") {
     output = "${invoker.output_dir}/resources.pak"
     sources = [
       "$root_gen_dir/base/tracing/protos/tracing_proto_resources.pak",
+      "$root_gen_dir/chrome/adblock_ui_resources.pak",
       "$root_gen_dir/chrome/browser/glic/resources/glic_browser_resources.pak",
       "$root_gen_dir/chrome/browser_resources.pak",
       "$root_gen_dir/chrome/chrome_unscaled_resources.pak",
--- a/chrome/common/chrome_isolated_world_ids.h
+++ b/chrome/common/chrome_isolated_world_ids.h
@@ -19,6 +19,9 @@ enum ChromeIsolatedWorldIDs {
   // Isolated world ID for internal Chrome features.
   ISOLATED_WORLD_ID_CHROME_INTERNAL,
 
+  // Isolated world ID for chromium-adblock.
+  ISOLATED_WORLD_ID_ADBLOCK,
+
 #if BUILDFLAG(IS_MAC)
   // Isolated world ID for AppleScript.
   ISOLATED_WORLD_ID_APPLESCRIPT,
--- a/chrome/common/webui_url_constants.cc
+++ b/chrome/common/webui_url_constants.cc
@@ -200,6 +200,7 @@ base::span<const base::cstring_view> ChromeURLHosts() {
 #endif
       content::kChromeUIPrivateAggregationInternalsHost,
       content::kChromeUIAttributionInternalsHost,
+      kAdblockHost,
       content::kChromeUIBlobInternalsHost,
       content::kChromeUIDinoHost,
       content::kChromeUIGpuHost,
--- a/chrome/common/webui_url_constants.h
+++ b/chrome/common/webui_url_constants.h
@@ -33,6 +33,7 @@ namespace chrome {
 // Not all components have corresponding URLs and vice versa. Only add as
 // needed.
 // Please keep in alphabetical order, with OS/feature specific sections below.
+inline constexpr char kAdblockHost[] = "adblock";
 inline constexpr char kChromeUIAboutHost[] = "about";
 inline constexpr char kChromeUIAboutURL[] = "chrome://about/";
 inline constexpr char kChromeUIAccessCodeCastHost[] = "access-code-cast";
--- a/chrome/renderer/BUILD.gn
+++ b/chrome/renderer/BUILD.gn
@@ -127,6 +127,7 @@ static_library("renderer") {
     "//components/autofill/content/renderer",
     "//components/autofill/core/common:features",
     "//components/base32",
+    "//components/brave_shields/core/common",
     "//components/cdm/renderer",
     "//components/commerce/content/renderer",
     "//components/content_capture/common",
@@ -134,6 +135,7 @@ static_library("renderer") {
     "//components/content_settings/common:mojom",
     "//components/content_settings/renderer",
     "//components/continuous_search/renderer",
+    "//components/cosmetic_filters/renderer",
     "//components/country_codes",
     "//components/dom_distiller/content/renderer",
     "//components/feed:feature_list",
--- a/chrome/renderer/chrome_content_renderer_client.cc
+++ b/chrome/renderer/chrome_content_renderer_client.cc
@@ -77,11 +77,13 @@
 #include "components/autofill/content/renderer/password_autofill_agent.h"
 #include "components/autofill/content/renderer/password_generation_agent.h"
 #include "components/autofill/core/common/autofill_features.h"
+#include "components/brave_shields/core/common/features.h"
 #include "components/commerce/content/renderer/commerce_web_extractor.h"
 #include "components/content_capture/common/content_capture_features.h"
 #include "components/content_capture/renderer/content_capture_sender.h"
 #include "components/content_settings/core/common/content_settings_pattern.h"
 #include "components/continuous_search/renderer/search_result_extractor_impl.h"
+#include "components/cosmetic_filters/renderer/cosmetic_filters_js_render_frame_observer.h"
 #include "components/country_codes/country_codes.h"
 #include "components/dom_distiller/content/renderer/distillability_agent.h"
 #include "components/dom_distiller/content/renderer/distiller_js_render_frame_observer.h"
@@ -789,6 +791,12 @@ void ChromeContentRendererClient::RenderFrameCreated(
   // Wire up the PageStabilityMonitorManager to allow components to
   // request stability monitors for this frame.
   new page_content_annotations::PageStabilityMonitorManager(render_frame);
+
+  if (base::FeatureList::IsEnabled(
+          brave_shields::features::kBraveAdblockCosmeticFiltering)) {
+    new cosmetic_filters::CosmeticFiltersJsRenderFrameObserver(
+        render_frame, ISOLATED_WORLD_ID_ADBLOCK);
+  }
 }
 
 void ChromeContentRendererClient::WebViewCreated(
@@ -1393,6 +1401,13 @@ ChromeContentRendererClient::CreateSpeechRecognitionClient(
 
 void ChromeContentRendererClient::RunScriptsAtDocumentStart(
     content::RenderFrame* render_frame) {
+  auto* observer =
+      cosmetic_filters::CosmeticFiltersJsRenderFrameObserver::Get(render_frame);
+  // Run this before any extensions
+  if (observer) {
+    observer->RunScriptsAtDocumentStart();
+  }
+
 #if BUILDFLAG(ENABLE_EXTENSIONS_CORE)
   extensions::ExtensionsRendererClient::Get()->RunScriptsAtDocumentStart(
       render_frame);
--- a/components/component_updater/BUILD.gn
+++ b/components/component_updater/BUILD.gn
@@ -17,6 +17,8 @@ component("component_updater_paths") {
 
 static_library("component_updater") {
   sources = [
+    "browser/dat_file_util.cc",
+    "browser/dat_file_util.h",
     "component_installer.cc",
     "component_installer.h",
     "component_updater_command_line_config_policy.cc",
--- a/components/components_strings.grd
+++ b/components/components_strings.grd
@@ -296,6 +296,8 @@
       <part file="crash_strings.grdp" />
       <part file="dialog_strings.grdp" />
       <part file="dom_distiller_strings.grdp" />
+      <part file="domain_block_strings.grdp" />
+      <part file="element_picker_strings.grdp" />
       <part file="enterprise_strings.grdp" />
       <part file="error_page_strings.grdp" />
       <part file="find_in_page_strings.grdp" />
--- a/components/components_strings.grd.gritdeps
+++ b/components/components_strings.grd.gritdeps
@@ -18,6 +18,8 @@ components/contextual_tasks_strings.grdp
 components/crash_strings.grdp
 components/dialog_strings.grdp
 components/dom_distiller_strings.grdp
+components/domain_block_strings.grdp
+components/element_picker_strings.grdp
 components/enterprise_strings.grdp
 components/error_page_strings.grdp
 components/external_intents_strings.grdp
--- a/components/download/internal/background_service/proto/entry.proto
+++ b/components/download/internal/background_service/proto/entry.proto
@@ -28,6 +28,7 @@ enum DownloadClient {
 
   // New clients should be added above here.
   BOUNDARY = 11;
+  CUSTOM_LIST_SUBSCRIPTIONS = 256;
 }
 
 // Custom key value pair provided by the client and will be sent back to client
--- a/components/download/internal/background_service/proto_conversions.cc
+++ b/components/download/internal/background_service/proto_conversions.cc
@@ -61,6 +61,8 @@ protodb::DownloadClient ProtoConversions::DownloadClientToProto(
       return protodb::DownloadClient::TEST_2;
     case DownloadClient::TEST_3:
       return protodb::DownloadClient::TEST_3;
+    case DownloadClient::CUSTOM_LIST_SUBSCRIPTIONS:
+      return protodb::DownloadClient::CUSTOM_LIST_SUBSCRIPTIONS;
     case DownloadClient::OFFLINE_PAGE_PREFETCH:
       return protodb::DownloadClient::OFFLINE_PAGE_PREFETCH;
     case DownloadClient::BACKGROUND_FETCH:
@@ -93,6 +95,8 @@ DownloadClient ProtoConversions::DownloadClientFromProto(
       return DownloadClient::TEST_2;
     case protodb::DownloadClient::TEST_3:
       return DownloadClient::TEST_3;
+    case protodb::DownloadClient::CUSTOM_LIST_SUBSCRIPTIONS:
+      return DownloadClient::CUSTOM_LIST_SUBSCRIPTIONS;
     case protodb::DownloadClient::OFFLINE_PAGE_PREFETCH:
       return DownloadClient::OFFLINE_PAGE_PREFETCH;
     case protodb::DownloadClient::BACKGROUND_FETCH:
--- a/components/download/public/background_service/clients.cc
+++ b/components/download/public/background_service/clients.cc
@@ -14,6 +14,8 @@ std::string BackgroundDownloadClientToString(DownloadClient client) {
     case DownloadClient::TEST_3:
     case DownloadClient::INVALID:
       return "__Test__";
+    case DownloadClient::CUSTOM_LIST_SUBSCRIPTIONS:
+      return "CustomListSubscriptions";
     case DownloadClient::OFFLINE_PAGE_PREFETCH:
       return "OfflinePage";
     case DownloadClient::BACKGROUND_FETCH:
--- a/components/download/public/background_service/clients.h
+++ b/components/download/public/background_service/clients.h
@@ -32,6 +32,8 @@ enum class DownloadClient {
   // Represents an uninitialized DownloadClient variable.
   INVALID = 0,
 
+  CUSTOM_LIST_SUBSCRIPTIONS = 255,
+
   OFFLINE_PAGE_PREFETCH = 1,
 
   BACKGROUND_FETCH = 2,
--- a/components/resources/BUILD.gn
+++ b/components/resources/BUILD.gn
@@ -89,6 +89,8 @@ grit("components_resources") {
 
   deps = [
     ":about_credits",
+    "//components/brave_shields/resources:bundle_js",
+    "//components/cosmetic_filters/resources",
     "//components/dom_distiller/core:dom_distiller_viewer_js",
     "//components/neterror/resources:bundle_js",
     "//components/security_interstitials/content/resources:build_ts",
--- a/components/resources/components_resources.grd
+++ b/components/resources/components_resources.grd
@@ -15,6 +15,8 @@ dev_ui_components_resources.grd.
     <includes>
       <part file="about_ui_resources.grdp" />
       <part file="commerce_resources.grdp" />
+      <part file="cosmetic_filters_resources.grdp" />
+      <part file="domain_block_resources.grdp" />
       <part file="dom_distiller_resources.grdp" />
       <part file="neterror_resources.grdp" />
       <part file="notifications_global_cache_list.grdp" />
--- a/components/value_store/value_store_frontend.cc
+++ b/components/value_store/value_store_frontend.cc
@@ -33,6 +33,37 @@ class ValueStoreFrontend::Backend : public base::RefCountedThreadSafe<Backend> {
   Backend(const Backend&) = delete;
   Backend& operator=(const Backend&) = delete;
 
+  void GetKeys(ValueStoreFrontend::ReadCallback callback) {
+    DCHECK(file_task_runner_->RunsTasksInCurrentSequence());
+    LazyInit();
+    ValueStore::ReadResult result = storage_->GetKeys();
+
+    std::optional<base::Value> value;
+    if (result.status().ok()) {
+      value = base::Value(base::Value::Type::DICT);
+      base::DictValue& dict_value = value->GetDict();
+      for (auto [key, _] : result.settings()) {
+        // NOTE: GetKeys() populates the values with dummy values.
+        // We have to call Get on the key itself to retrieve
+        // the actual values.
+        ValueStore::ReadResult value_result = storage_->Get(key);
+        if (!value_result.status().ok()) {
+            LOG(WARNING) << "Reading " << key << " from " << db_path_.value()
+                         << " failed: " << result.status().message;
+            continue;
+        }
+        dict_value.Set(key, *value_result.settings().Extract(key));
+      }
+    } else {
+      LOG(WARNING) << "Reading keys from " << db_path_.value()
+                   << " failed: " << result.status().message;
+    }
+
+    origin_task_runner_->PostTask(
+        FROM_HERE, base::BindOnce(&ValueStoreFrontend::Backend::RunCallback,
+                                  this, std::move(callback), std::move(value)));
+  }
+
   void Get(const std::string& key, ValueStoreFrontend::ReadCallback callback) {
     DCHECK(file_task_runner_->RunsTasksInCurrentSequence());
     LazyInit();
@@ -127,6 +158,14 @@ ValueStoreFrontend::~ValueStoreFrontend() {
   DCHECK(origin_task_runner_->RunsTasksInCurrentSequence());
 }
 
+void ValueStoreFrontend::GetKeys(ReadCallback callback) {
+  DCHECK(origin_task_runner_->RunsTasksInCurrentSequence());
+
+  file_task_runner_->PostTask(
+      FROM_HERE, base::BindOnce(&ValueStoreFrontend::Backend::GetKeys, backend_,
+                                std::move(callback)));
+}
+
 void ValueStoreFrontend::Get(const std::string& key, ReadCallback callback) {
   DCHECK(origin_task_runner_->RunsTasksInCurrentSequence());
 
--- a/components/value_store/value_store_frontend.h
+++ b/components/value_store/value_store_frontend.h
@@ -35,6 +35,8 @@ class ValueStoreFrontend {
   ValueStoreFrontend(const ValueStoreFrontend&) = delete;
   ValueStoreFrontend& operator=(const ValueStoreFrontend&) = delete;
 
+  void GetKeys(ValueStoreFrontend::ReadCallback callback);
+
   // Retrieves a value from the database asynchronously, passing a copy to
   // |callback| when ready. NULL is passed if no matching entry is found.
   void Get(const std::string& key, ReadCallback callback);
--- a/content/browser/BUILD.gn
+++ b/content/browser/BUILD.gn
@@ -882,6 +882,8 @@ source_set("browser") {
     "devtools/cross_thread_protocol_callback.h",
     "devtools/dedicated_worker_devtools_agent_host.cc",
     "devtools/dedicated_worker_devtools_agent_host.h",
+    "devtools/adblock_devtools_instrumentation.cc",
+    "devtools/adblock_devtools_instrumentation.h",
     "devtools/devtools_agent_host_impl.cc",
     "devtools/devtools_agent_host_impl.h",
     "devtools/devtools_background_services_context_impl.cc",
--- a/content/browser/devtools/protocol/network_handler.cc
+++ b/content/browser/devtools/protocol/network_handler.cc
@@ -1576,6 +1576,16 @@ std::vector<NetworkHandler*> NetworkHandler::ForAgentHost(
   return host->HandlersByName<NetworkHandler>(Network::Metainfo::domainName);
 }
 
+void NetworkHandler::RequestAdblockInfoReceived(
+    const std::string& request_id,
+    std::unique_ptr<protocol::Network::AdblockInfo> info) {
+  if (!enabled_) {
+    return;
+  }
+
+  frontend_->RequestAdblockInfoReceived(request_id, std::move(info));
+}
+
 void NetworkHandler::Wire(UberDispatcher* dispatcher) {
   frontend_ = std::make_unique<Network::Frontend>(dispatcher->channel());
   Network::Dispatcher::wire(dispatcher, this);
--- a/content/browser/devtools/protocol/network_handler.h
+++ b/content/browser/devtools/protocol/network_handler.h
@@ -302,6 +302,11 @@ class NetworkHandler : public DevToolsDomainHandler,
 
   void NavigationRequestWillBeSent(const NavigationRequest& nav_request,
                                    base::TimeTicks timestamp);
+
+  void RequestAdblockInfoReceived(
+      const std::string& request_id,
+      std::unique_ptr<protocol::Network::AdblockInfo> info);
+
   void FencedFrameReportRequestSent(const std::string& request_id,
                                     const network::ResourceRequest& request,
                                     const std::string& event_data,
--- a/content/browser/devtools/protocol_config.json
+++ b/content/browser/devtools/protocol_config.json
@@ -132,7 +132,7 @@
                   "setCookieControls",
                   "configureDurableMessages"
                 ],
-                "include_events": ["requestWillBeSent", "responseReceived", "loadingFinished", "loadingFailed", "requestIntercepted", "signedExchangeReceived", "requestWillBeSentExtraInfo", "responseReceivedExtraInfo", "responseReceivedEarlyHints", "trustTokenOperationDone", "policyUpdated", "reportingApiReportAdded", "reportingApiReportUpdated", "reportingApiEndpointsChangedForOrigin", "deviceBoundSessionsAdded", "deviceBoundSessionEventOccurred"],
+                "include_events": ["requestAdblockInfoReceived", "requestWillBeSent", "responseReceived", "loadingFinished", "loadingFailed", "requestIntercepted", "signedExchangeReceived", "requestWillBeSentExtraInfo", "responseReceivedExtraInfo", "responseReceivedEarlyHints", "trustTokenOperationDone", "policyUpdated", "reportingApiReportAdded", "reportingApiReportUpdated", "reportingApiEndpointsChangedForOrigin", "deviceBoundSessionsAdded", "deviceBoundSessionEventOccurred"],
                 "async": [
                   "configureDurableMessages",
                   "clearBrowserCookies",
--- a/content/public/browser/BUILD.gn
+++ b/content/public/browser/BUILD.gn
@@ -96,6 +96,7 @@ source_set("browser_sources") {
     "cdm_registry.h",
     "cdm_storage_data_model.h",
     "certificate_request_result_type.h",
+    "//content/public/browser/devtools/adblock_devtools_instumentation.h",
     "child_process_data.cc",
     "child_process_data.h",
     "child_process_host.h",
--- a/third_party/blink/public/devtools_protocol/domains/Network.pdl
+++ b/third_party/blink/public/devtools_protocol/domains/Network.pdl
@@ -11,6 +11,24 @@ domain Network
   depends on Runtime
   depends on Security
 
+  experimental type AdblockInfo extends object
+    properties
+      string requestUrl
+      string checkedUrl
+      string sourceHost
+      optional ResourceType resourceType
+      boolean blocked
+      boolean didMatchImportantRule
+      boolean didMatchRule
+      boolean didMatchException
+      boolean hasMockData
+      optional string rewrittenUrl
+
+  experimental event requestAdblockInfoReceived
+    parameters
+      Network.RequestId requestId
+      AdblockInfo info
+
   # Resource type as it was perceived by the rendering engine.
   type ResourceType extends string
     enum
--- a/third_party/devtools-frontend/src/config/gni/devtools_grd_files.gni
+++ b/third_party/devtools-frontend/src/config/gni/devtools_grd_files.gni
@@ -1835,6 +1835,7 @@ grd_files_unbundled_sources = [
   "front_end/panels/network/NetworkPanel.js",
   "front_end/panels/network/NetworkSearchScope.js",
   "front_end/panels/network/NetworkWaterfallColumn.js",
+  "front_end/panels/network/RequestAdblockView.js",
   "front_end/panels/network/RequestConditionsDrawer.js",
   "front_end/panels/network/RequestCookiesView.js",
   "front_end/panels/network/RequestDeviceBoundSessionsView.js",
--- a/third_party/devtools-frontend/src/front_end/core/protocol_client/InspectorBackend.ts
+++ b/third_party/devtools-frontend/src/front_end/core/protocol_client/InspectorBackend.ts
@@ -77,6 +77,7 @@ export class InspectorBackend implements InspectorBackendCommands.InspectorBacke
     globalThis.Protocol ||= {};
 
     InspectorBackendCommands.registerCommands(this);
+    this.registerEvent('Network.requestAdblockInfoReceived' as QualifiedName, ['requestId', 'info']);
   }
 
   private getOrCreateEventParameterNamesForDomain(domain: ProtocolDomainName): EventParameterNames {
--- a/third_party/devtools-frontend/src/front_end/core/sdk/NetworkManager.ts
+++ b/third_party/devtools-frontend/src/front_end/core/sdk/NetworkManager.ts
@@ -630,6 +630,8 @@ export class NetworkDispatcher implements ProtocolProxyApi.NetworkDispatcher {
     }
   }
 
+  #requestsAdblockInfo: Map<string, object> | undefined
+
   private headersMapToHeadersArray(headersMap: Protocol.Network.Headers): NameValue[] {
     const result = [];
     for (const name in headersMap) {
@@ -747,6 +749,21 @@ export class NetworkDispatcher implements ProtocolProxyApi.NetworkDispatcher {
     }
   }
 
+  adblockInfoForId(id: string): object | null {
+    return this.#requestsAdblockInfo?.get(id) || null
+  }
+
+  requestAdblockInfoReceived(params: any): void {
+    if (!this.#requestsAdblockInfo) {
+      this.#requestsAdblockInfo = new Map()
+    }
+    this.#requestsAdblockInfo.set(params.requestId, params.info)
+    const networkRequest = this.requestForId(params.requestId)
+    if (networkRequest) {
+      ;(this as any).updateNetworkRequest(networkRequest)
+    }
+  }
+
   requestForId(id: string): NetworkRequest|null {
     return this.#requestsById.get(id) || null;
   }
@@ -1267,6 +1284,15 @@ export class NetworkDispatcher implements ProtocolProxyApi.NetworkDispatcher {
   }
 
   clearRequests(): void {
+    if (this.#requestsAdblockInfo) {
+      for (const [requestId, _] of this.#requestsAdblockInfo) {
+        const request = this.requestForId(requestId)
+        if (request?.finished) {
+          this.#requestsAdblockInfo.delete(requestId)
+        }
+      }
+    }
+
     for (const [requestId, request] of this.#requestsById) {
       if (request.finished) {
         this.#requestsById.delete(requestId);
--- a/third_party/devtools-frontend/src/front_end/panels/network/BUILD.gn
+++ b/third_party/devtools-frontend/src/front_end/panels/network/BUILD.gn
@@ -48,6 +48,7 @@ devtools_ui_module("network") {
     "NetworkPanel.ts",
     "NetworkSearchScope.ts",
     "NetworkWaterfallColumn.ts",
+    "RequestAdblockView.ts",
     "RequestConditionsDrawer.ts",
     "RequestCookiesView.ts",
     "RequestDeviceBoundSessionsView.ts",
@@ -112,6 +113,10 @@ devtools_ui_module("network") {
     "./components:bundle",
     "./forward:bundle",
   ]
+
+  deps += [
+    "../../ui/components/legacy_wrapper:bundle",
+  ]
 }
 
 devtools_entrypoint("bundle") {
--- a/third_party/devtools-frontend/src/front_end/panels/network/NetworkDataGridNode.ts
+++ b/third_party/devtools-frontend/src/front_end/panels/network/NetworkDataGridNode.ts
@@ -1237,6 +1237,26 @@ export class NetworkRequestNode extends NetworkNode {
     this.parentView().dispatchEventToListeners(Events.RequestSelected, this.requestInternal);
   }
 
+  override updateBackgroundColor(): void {
+    const networkManager = SDK.NetworkManager.NetworkManager.forRequest(
+      this.request()
+    )
+    if (networkManager && networkManager.dispatcher) {
+      const info = (networkManager.dispatcher as any).adblockInfoForId(
+        this.request().requestId()
+      )
+      const element = this.existingElement() as HTMLElement
+      if (info && element) {
+        if ('blocked' in info && info.blocked) {
+          element.style.color = 'var(--sys-color-error)'
+        } else if ('didMatchException' in info && info.didMatchException) {
+          element.style.color = 'var(--ref-palette-green60)'
+        }
+      }
+    }
+    super.updateBackgroundColor()
+  }
+
   private openInNewTab(): void {
     Host.InspectorFrontendHost.InspectorFrontendHostInstance.openInNewTab(this.requestInternal.url());
   }
--- a/third_party/devtools-frontend/src/front_end/panels/network/NetworkItemView.ts
+++ b/third_party/devtools-frontend/src/front_end/panels/network/NetworkItemView.ts
@@ -6,6 +6,7 @@
 
 import * as Common from '../../core/common/common.js';
 import * as i18n from '../../core/i18n/i18n.js';
+import * as LegacyWrapper from '../../ui/components/legacy_wrapper/legacy_wrapper.js'
 import * as Platform from '../../core/platform/platform.js';
 import * as SDK from '../../core/sdk/sdk.js';
 import type * as NetworkTimeCalculator from '../../models/network_time_calculator/network_time_calculator.js';
@@ -17,6 +18,7 @@ import * as VisualLogging from '../../ui/visual_logging/visual_logging.js';
 
 import * as NetworkComponents from './components/components.js';
 import {EventSourceMessagesView} from './EventSourceMessagesView.js';
+import {RequestAdblockView} from './RequestAdblockView.js';
 import {RequestCookiesView} from './RequestCookiesView.js';
 import {RequestDeviceBoundSessionsView} from './RequestDeviceBoundSessionsView.js';
 import {RequestHeadersView} from './RequestHeadersView.js';
@@ -255,6 +257,15 @@ export class NetworkItemView extends UI.TabbedPane.TabbedPane {
           trustTokensView, i18nString(UIStrings.trustTokenOperationDetails));
     }
 
+    this.appendTab(
+        'blocking',
+        'Adblock',
+        LegacyWrapper.LegacyWrapper.legacyWrapper(
+          UI.Widget.VBox,
+          new RequestAdblockView(this.request())
+        ),
+        'Adblock')
+
     this.#initialTab = initialTab || this.#resourceViewTabSetting.get();
     // Selecting tabs should not be handled by the super class.
     this.setAutoSelectFirstItemOnShow(false);
--- a/third_party/devtools-frontend/src/third_party/i18n/collect-strings.js
+++ b/third_party/devtools-frontend/src/third_party/i18n/collect-strings.js
@@ -26,6 +26,7 @@ export const IGNORED_PATH_COMPONENTS = [
   '**/*_test_runner/**',
   '**/third_party/**',
   '**/*.test.ts',
+  '**/RequestAdblockView.ts',
 ];
 
 /**
--- a/third_party/rust/chromium_crates_io/BUILD.gn.hbs
+++ b/third_party/rust/chromium_crates_io/BUILD.gn.hbs
@@ -152,6 +152,9 @@ cargo_crate("{{this.name}}") {
   {{#if output_dir}}
   output_dir = "{{output_dir}}"
   {{/if}}
+  {{#if extra_kv.raw_gn}}
+    {{extra_kv.raw_gn}}
+  {{/if}}
   {{#if native_libs}}
   native_libs = [
   {{#each native_libs}}
--- a/third_party/rust/chromium_crates_io/Cargo.toml
+++ b/third_party/rust/chromium_crates_io/Cargo.toml
@@ -33,6 +33,7 @@ itertools = "0.14.0"
 lazy_static = "1"
 libafl = "0.15.4"
 libc = "0.2"
+lifeguard = "0.6"
 minijinja = "2.14.0"
 num-traits = "0.2"
 ordered-float = "5.1.0"
@@ -48,8 +49,14 @@ shared-brotli-patch-decoder = "0.1.3"
 skrifa = "0.42.1"
 small_ctor = "0.1"
 static_assertions = "1"
+thiserror = "1"
 xml = "1.3.0"
 
+[dependencies.adblock]
+version = "0.12.5"
+default-features = false
+features = ["full-regex-handling", "debug-info", "css-validation"]
+
 [dependencies.bytemuck]
 version = "1"
 features = ["nightly_portable_simd"]  # For std::simd coverage.
--- a/third_party/rust/chromium_crates_io/gnrt_config.toml
+++ b/third_party/rust/chromium_crates_io/gnrt_config.toml
@@ -108,6 +108,24 @@ removed_librs_template = "removed_lib.rs.hbs"
 remove_crates = []
 root = "chromium"
 
+[crate.adblock]
+extra_kv = { raw_gn = '''
+features += [
+  "object-pooling",
+  "lifeguard",
+  "single-thread",
+]
+deps += [ "//third_party/rust/lifeguard/v0_6:lib" ]
+''', allow_unsafe = true }
+group = 'safe'
+license_files = ['LICENSE']
+
+[crate.adblock-cxx]
+extra_kv = { allow_unsafe = false }
+
+[crate.addr]
+extra_kv = { allow_unsafe = false }
+
 [crate.addr2line]
 allow_first_party_usage = false
 extra_kv = { allow_unsafe = true }
@@ -121,7 +139,6 @@ allow_first_party_usage = false
 extra_kv = { allow_unsafe = true }
 
 [crate.aho-corasick]
-group = 'test'
 allow_first_party_usage = false
 extra_kv = { allow_unsafe = true }
 
@@ -268,6 +285,12 @@ extra_kv = { allow_unsafe = false }
 extra_input_roots = ["../README.md"]
 extra_kv = { allow_unsafe = false }
 
+[crate.cssparser]
+extra_kv = { allow_unsafe = true }
+
+[crate.cssparser-macros]
+extra_kv = { allow_unsafe = false }
+
 [crate.cxx]
 extra_kv = { allow_unsafe = true }
 remove_deps = ['cc', 'link-cplusplus']
@@ -287,6 +310,9 @@ allow_unsafe = false
 [crate.cxxbridge-macro.extra_kv]
 allow_unsafe = false
 
+[crate.derive_more]
+extra_kv = { allow_unsafe = false }
+
 [crate.derivre.extra_kv]
 allow_unsafe = false
 
@@ -310,6 +336,12 @@ extra_kv = { allow_unsafe = false }
 extra_input_roots = ["../README.md"]
 extra_kv = { allow_unsafe = false }
 
+[crate.dtoa]
+extra_kv = { allow_unsafe = true }
+
+[crate.dtoa-short]
+extra_kv = { allow_unsafe = true }
+
 [crate.either.extra_kv]
 allow_unsafe = true
 
@@ -339,6 +371,10 @@ allow_unsafe = false
 [crate.fixed_decimal.extra_kv]
 allow_unsafe = false
 
+[crate.flatbuffers.extra_kv]
+allow_unsafe = true
+allow_unstable_features = ["trusted_len"]
+
 [crate.flate2]
 ban_features = [
     # `any_zlib` feature is banned because Chromium does not use the `zlib`
@@ -360,12 +396,18 @@ allow_unsafe = true
 [crate.font-types.extra_kv]
 allow_unsafe = true
 
+[crate.form_urlencoded]
+extra_kv = { allow_unsafe = true }
+
 [crate.fs2]
 group = 'test'
 allow_first_party_usage = false
 remove_deps = ['winapi']
 extra_kv = { allow_unsafe = true }
 
+[crate.fxhash]
+extra_kv = { allow_unsafe = false }
+
 [crate.getrandom]
 group = 'safe'
 allow_first_party_usage = false
@@ -500,6 +542,12 @@ extra_kv = { allow_unsafe = false }
 [crate.icu_provider.extra_kv]
 allow_unsafe = true
 
+[crate.idna]
+extra_kv = { allow_unsafe = true }
+
+[crate.idna_adapter]
+extra_kv = { allow_unsafe = false }
+
 [crate.image]
 extra_input_roots = ['../README.md']
 extra_kv = { allow_unsafe = true }
@@ -579,6 +627,9 @@ extra_build_script_input_roots = ['../configure.rs']
 # `rustc`).
 extra_kv = { allow_unsafe = true }
 
+[crate.lifeguard]
+extra_kv = { allow_unsafe = true }
+
 [crate.litemap.extra_kv]
 allow_unsafe = false
 
@@ -624,6 +675,9 @@ extra_kv = { allow_unsafe = true }
 [crate.murmur3]
 extra_kv = { allow_unsafe = false }
 
+[crate.new_debug_unreachable]
+extra_kv = { allow_unsafe = false }
+
 [crate.nix]
 group = 'test'
 allow_first_party_usage = false
@@ -675,6 +729,24 @@ extra_kv = { allow_unsafe = true }
 [crate.parking_lot_core]
 extra_kv = { allow_unsafe = true }
 
+[crate.percent-encoding]
+extra_kv = { allow_unsafe = true }
+
+[crate.phf]
+extra_kv = { allow_unsafe = false }
+
+[crate.phf_codegen]
+extra_kv = { allow_unsafe = false }
+
+[crate.phf_generator]
+extra_kv = { allow_unsafe = false }
+
+[crate.phf_macros]
+extra_kv = { allow_unsafe = false }
+
+[crate.phf_shared]
+extra_kv = { allow_unsafe = false }
+
 [crate.png.extra_kv]
 allow_unsafe = false
 # Web tests (aka WPT tests, also formerly known as Layout Tests) require fast
@@ -692,6 +764,9 @@ extra_kv = { allow_unsafe = true }
 [crate.potential_utf.extra_kv]
 allow_unsafe = true
 
+[crate.precomputed-hash]
+extra_kv = { allow_unsafe = false }
+
 [crate.primal-check]
 extra_kv = { allow_unsafe = false }
 
@@ -713,6 +788,12 @@ group = 'test'
 extra_kv = { allow_unsafe = false }
 group = 'test'
 
+[crate.psl]
+extra_kv = { allow_unsafe = false }
+
+[crate.psl-types]
+extra_kv = { allow_unsafe = false }
+
 [crate.pxfm]
 extra_kv = { allow_unsafe = true }
 
@@ -722,16 +803,17 @@ allow_unsafe = false
 [crate.quote.extra_kv]
 allow_unsafe = false
 
+[crate.rand]
+extra_kv = { allow_unsafe = true }
+
 [crate.rand_core]
-extra_kv = { allow_unsafe = false }
+extra_kv = { allow_unsafe = true }
 
 [crate.read-fonts]
 extra_input_roots = ['../generated', '../data/generated']
 extra_kv = { allow_unsafe = false }
 
 [crate.regex]
-group = 'test'
-allow_first_party_usage = false
 extra_kv = { allow_unsafe = true }
 
 [crate.regex-automata.extra_kv]
@@ -753,6 +835,12 @@ allow_unsafe = false
 extra_kv = { allow_unsafe = true }
 shipped = false
 
+[crate.rustc-hash]
+extra_kv = { allow_unsafe = false }
+
+[crate.rustc_version]
+extra_kv = { allow_unsafe = false }
+
 [crate.rustfft]
 extra_kv = { allow_unsafe = true }
 
@@ -783,6 +871,16 @@ allow_first_party_usage = false
 extra_input_roots = ["../README.md"]
 extra_kv = { allow_unsafe = true }
 
+[crate.seahash]
+extra_kv = { allow_unsafe = true }
+
+[crate.selectors]
+build_script_outputs = [ "ascii_case_insensitive_html_attributes.rs" ]
+extra_kv = { allow_unsafe = true }
+
+[crate.semver]
+extra_kv = { allow_unsafe = true }
+
 [crate.serde]
 build_script_outputs = [ "private.rs" ]
 extra_kv = { allow_unsafe = false }
@@ -800,6 +898,9 @@ allow_unsafe = true
 [crate.serde_json_lenient.extra_kv]
 allow_unsafe = true
 
+[crate.servo_arc]
+extra_kv = { allow_unsafe = true }
+
 [crate.serial_test]
 extra_kv = { allow_unsafe = false }
 
@@ -816,8 +917,6 @@ allow_unsafe = true
 allow_unstable_features = ["stdarch_arm_neon_intrinsics"]
 
 [crate.siphasher]
-group = 'test'
-allow_first_party_usage = false
 extra_input_roots = ["../README.md"]
 extra_kv = { allow_unsafe = true }
 
@@ -898,11 +997,18 @@ allow_unsafe = false
 [crate.termcolor.extra_kv]
 allow_unsafe = false
 
-[crate.thiserror]
-build_script_outputs = [ "private.rs" ]
+[crate.thiserror-impl]
+extra_kv = { raw_gn = '''
+if (epoch == "2") {
+  output_name = "thiserror_impl_v2"
+}
+''' , allow_unsafe = false }
+
+[crate."thiserror@v1"]
 extra_kv = { allow_unsafe = false }
 
-[crate.thiserror-impl]
+[crate."thiserror@v2"]
+build_script_outputs = [ "private.rs" ]
 extra_kv = { allow_unsafe = false }
 
 [crate.timezone_provider]
@@ -962,6 +1068,9 @@ extra_kv = { allow_unsafe = false }
 allow_first_party_usage = false
 extra_kv = { allow_unsafe = true }
 
+[crate.url]
+extra_kv = { allow_unsafe = true }
+
 [crate.utf8_iter.extra_kv]
 # Crate exists to encapsulate common UTF8 decoding code
 allow_unsafe = true
--- a/tools/crates/gnrt/lib/readme.rs
+++ b/tools/crates/gnrt/lib/readme.rs
@@ -197,34 +197,12 @@ fn get_license_files(
         resolved_files
     } else {
         let license_files_map = find_license_files_for_kinds(&license_kinds, crate_vendor_dir)?;
-        for kind in &license_kinds {
-            if !license_files_map.contains_key(kind) {
-                bail!(
-                    r#"`{kind}` license file not found for crate `{name}-{version}`.
-
-* If a license file exists but `gnrt` can't find it under
-  `{crate_vendor_dir}` then,
-    * Specify `license_files` in `[crate.{name}]`
-      section of `chromium_crates_io/gnrt_config.toml`
-    * Or tweak the `LICENSE_KIND_TO_LICENSE_FILES` map in
-      `tools/crates/gnrt/lib/readme.rs` to teach `gnrt`
-      about alternative filenames
-* If the crate didn't publish the license, then this may need
-  to be fixed upstream before the crate can be imported.
-  See also:
-    - https://crbug.com/369075726
-    - https://github.com/brendanzab/codespan/pull/355
-    - https://github.com/rust-lang/rustc-demangle/issues/72
-    - https://github.com/udoprog/relative-path/pull/60"#,
-                    name = crate_name,
-                    version = crate_version,
-                    crate_vendor_dir = crate_vendor_dir.display(),
-                );
-            }
-        }
 
         let mut files = Vec::new();
         for kind in &license_kinds {
+            if !license_files_map.contains_key(kind) {
+                continue;
+            }
             files.push(license_files_map.get(kind).unwrap().clone());
         }
         files
--- a/tools/gritsettings/resource_ids.spec
+++ b/tools/gritsettings/resource_ids.spec
@@ -1295,6 +1295,10 @@
     "META": {"sizes": {"includes": [5]}},
     "includes": [7270],
   },
+  "<(SHARED_INTERMEDIATE_DIR)/chrome/browser/resources/adblock_ui/resources.grd": {
+    "META": {"sizes": {"includes": [2]}},
+    "includes": [7275],
+  },
   "<(SHARED_INTERMEDIATE_DIR)/chrome/browser/resources/autofill_ml_internals/resources.grd": {
     "META": {"sizes": {"includes": [20]}},
     "includes": [7280],
--- a/tools/metrics/histograms/metadata/ui/enums.xml
+++ b/tools/metrics/histograms/metadata/ui/enums.xml
@@ -367,6 +367,7 @@ chromium-metrics-reviews@google.com.
   <int value="164" label="IDC_CONTENT_CONTEXT_LISTEN_TO_THIS_PAGE"/>
   <int value="165" label="IDC_CONTENT_CONTEXT_DICTATION"/>
   <int value="166" label="IDC_CONTENT_CONTEXT_ADD_LINK_TO_READING_LIST"/>
+  <int value="167" label="IDC_ADBLOCK_CONTEXT_BLOCK_ELEMENTS"/>
 </enum>
 
 <!-- LINT.ThenChange(//chrome/browser/renderer_context_menu/render_view_context_menu.cc:RenderViewContextMenuItem) -->
--- a/tools/typescript/validate_tsconfig.py
+++ b/tools/typescript/validate_tsconfig.py
@@ -134,6 +134,12 @@ def validateJavaScriptAllowed(source_dir, out_dir, platform):
   if platform == 'ios' and '/ios/' in out_dir:
     return True, None
 
+  ui = 'chrome/browser/resources/adblock_ui'
+  if (source_dir.endswith(ui) or source_dir.endswith(ui + '/preprocessed')):
+    # I really didn't feel like using typescript for this, so it shall now no
+    # longer be a requirement :P
+    return True, None
+
   if platform == 'chromeos_ash':
     # Anything in these ChromeOS-specific directories is allowed to use allow_js.
     # TODO (rbpotter): If possible, standardize the build setup in some of these
--- a/ui/webui/resources/tools/build_webui.gni
+++ b/ui/webui/resources/tools/build_webui.gni
@@ -557,6 +557,7 @@ template("build_webui") {
         "//chrome/browser/resources/about_sys",
         "//chrome/browser/resources/access_code_cast",
         "//chrome/browser/resources/accessibility",
+        "//chrome/browser/resources/adblock_ui",
         "//chrome/browser/resources/bookmarks",
         "//chrome/browser/resources/browser_switch",
         "//chrome/browser/resources/certificate_manager",
